sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -477,3 +500,203 @@ aes_error_t aes_gcm_gctr(const aes_key_len_t key_len, uint32_t *key_shares[2],
 
   return kAesOk;
 }
+
+/**
+ * Verify that the lengths of AES-GCM parameters are acceptable.
+ *
+ * This routine can be used for both authenticated encryption and authenticated
+ * decryption; the lengths of the plaintext and ciphertext always match, and
+ * `plaintext_len` may represent either.
+ *
+ * @param iv_len IV length in bytes
+ * @param plaintext_len Plaintext/ciphertext length in bytes
+ * @param aad_len Associated data length in bytes
+ */
+static hardened_bool_t check_buffer_lengths(const size_t iv_len,
+                                            const size_t plaintext_len,
+                                            const size_t aad_len) {
+  // Check IV length (must be 96 or 128 bits = 12 or 16 bytes).
+  if (iv_len != 12 && iv_len != 16) {
+    return kHardenedBoolFalse;
+  }
+
+  // Check plaintext/AAD length. Both must be less than 2^32 bytes long. This
+  // is stricter than NIST requires, but SP800-38D also allows implementations
+  // to stipulate lower length limits.
+  if (plaintext_len > UINT32_MAX || aad_len > UINT32_MAX) {
+    return kHardenedBoolFalse;
+  }
+
+  return kHardenedBoolTrue;
+}
+
+/**
+ * Compute the hash subkey for AES-GCM.
+ *
+ * This routine computes the hash subkey H and the product table for H; see
+ * `make_product_table` for representation details.
+ *
+ * If any step in this process fails, the function returns an error and the
+ * output should not be used.
+ *
+ * @param key_len length of key
+ * @param key_shares key, expressed in two shares
+ * @param tbl Destination for the output hash subkey product table
+ * @return OK or error
+ */
+static aes_error_t aes_gcm_hash_subkey(const aes_key_len_t key_len,
+                                       uint32_t *key_shares[2],
```

Comment:
nit: Can this be `const uint32_t *key_shares[2]`.  There is also the option of refactoring `aes_params_t` so that `key` and `key_len` are provided via a separate struct.